### PR TITLE
Add FromStr impl for NonZero types

### LINKED ISSUE #58604: impl FromStr for NonZeroU8, NonZeroU16, ...
Is there any reason why the following types couldn't implement `core::str::FromStr`?

* `core::num::NonZeroU8`
* `core::num::NonZeroU16`
* `core::num::NonZeroU32`
* `core::num::NonZeroUsize`
* `core::num::NonZeroU64`
* `core::num::NonZeroU128`

### PR DESCRIPTION
**Pull request by hellow554**
This is a WIP implementation because I do have some questions regarding the solution.

Somebody should ping the lang team on this I guess.
Please see the annotations on the code for more details.

Closes #58604

### CODE REVIEW

**Comment by hellow554** [comment]
Just as a note: I'm on vacation the next two weeks, so I would really love to see some feedback soon @SimonSapin

**Comment by SimonSapin** [comment]
`#[unstable]` unfortunately doesn’t work on trait impls. Since both the type and trait are stable, the impl will be immediately usable.

@rfcbot fcp merge

> Please see the annotations on the code for more details.

I couldn’t find those.

**Comment by SimonSapin** [comment]
@rfcbot fcp merge

File: src/libcore/num/mod.rs
Comment (sfackler) [inline src/libcore/num/mod.rs]: There should be an accompanying Error implementation in libstd, right?

**Comment by hellow554** [comment]
@SimonSapin I mean the github review thing regarding the unstable thing and `from_str_radix`.

So for the unstable attribute, I can change that to stable with version 1.34?
What about `from_str_radix`?

**Comment by SimonSapin** [comment]
I don’t seem to find any review or comment about `from_str_radix` except this one 7 minutes ago. Maybe you still need to submit it?

I think that Beta 1.34 was or will be branched this week, so this should likely be 1.35. And yes, changed to stable.

**Comment by hellow554** [comment]
TIL: you don't see that rewiew until I hit that button ^^ Sorry!

**Comment by SimonSapin** [comment]
@rust-lang/libs What do you think of adding a `Zero` variant to `IntErrorKind` (which is `#[non_exhaustive]` *and* unstable) and reuse it and `ParseIntError` for these conversions, rather than introducing new types?

File: src/libcore/num/mod.rs
Comment (SimonSapin) [reply src/libcore/num/mod.rs]: This means something similar to this: (which is why the `__description` method is public.)

https://github.com/rust-lang/rust/blob/9fda7c2237db910e41d6a712e9a2139b352e558b/src/libstd/error.rs#L474-L479

File: src/libcore/num/mod.rs
Comment (SimonSapin) [reply src/libcore/num/mod.rs]: These `FromStr` impls are already for convenience rather than strictly necessary. I think parsing non-10 bases for non-zero types gets niche enough that it’s probably not worth dedicated API surface. You can always use e.g. `u32::from_str_radix` then `NonZeroU32::new`.

File: src/libcore/num/mod.rs
Comment (SimonSapin) [reply src/libcore/num/mod.rs]: Yes a feature name is needed although it’s basically invisible for immediately-stable features like this `impl`. I think the error type will need to be stable as well, but that’s not an issue if we reuse `ParseIntError`.

**Comment by alexcrichton** [comment]
@SimonSapin that strategy sounds plausible to me!

File: src/libcore/num/mod.rs
Comment (hellow554) [reply src/libcore/num/mod.rs]: Okay, I was wondering, because afaik the `ParseIntError` doesn't implement `Error`

Comment (SimonSapin) [reply src/libcore/num/mod.rs]: It… does? Exactly where I linked, in the comment you’re responding to.

**Comment by SimonSapin** [comment]
Alright, diff look good. r+ when FCP finishes.

File: src/libcore/num/mod.rs
Comment (hellow554) [reply src/libcore/num/mod.rs]: wow. I'm blind ^^ sorry

**Comment by hellow554** [comment]
```none
[01:25:14] error[E0658]: use of unstable library feature 'int_error_matching': it can be useful to match errors when making error messages for integer parsing (see issue #22639)
[01:25:14]  --> src/libcore/../libcore/tests/nonzero.rs:1:17
[01:25:14]   |
[01:25:14] 1 | use core::num::{IntErrorKind, NonZeroU8, NonZeroU32, NonZeroI32, ParseIntError};
[01:25:14]   |
[01:25:14]   = help: add #![feature(int_error_matching)] to the crate attributes to enable
```

How should I deal with this? Should I add the feature to the test or should I do it like https://github.com/rust-lang/rust/blob/master/src/libcore/tests/num/int_macros.rs#L158-L160 and completly ignore the error and just check if it's `Ok` or `Err`?

**Comment by SimonSapin** [comment]
I think the tests you linked were written before the `kind` method was added. So it’s not to avoid the feature gate that they are written this way. It’s fine to add whatever unstable feature gate to tests in this repository, as needed to write tests that you think are appropriate.

**Comment by hellow554** [comment]
Finally (!) fixed my local rust environment so I could run the tests again sucesfully. Cross your fingers!

**Comment by hellow554** [comment]
@SimonSapin merge is done. Can you do a r+?

**Comment by TimNN** [comment]
Ping from triage @SimonSapin: This PR is waiting for your review / approval!

**Comment by Centril** [comment]
Randomly reassigning; r? @Amanieu

**Comment by cuviper** [comment]
@bors r-

I bisected to this PR for the ICE in https://github.com/rust-lang/rust/pull/59466#issuecomment-477360976

```
[00:47:10] thread 'rustc' panicked at 'attempt to shift left with overflow', src/librustc_codegen_llvm/debuginfo/metadata.rs:1371:50
```

Which is from this line added in #55701:
https://github.com/rust-lang/rust/blob/33ef0bad21d6bb646c7c3ab0dbf381ca96c324bf/src/librustc_codegen_llvm/debuginfo/metadata.rs#L1371

I doubt this is your *fault*, per se, but it will need to be fixed for this to land. It also makes some sense that this wasn't triggered by the `x86_64-gnu-llvm-6.0` PR check, because older LLVM uses a fallback enum representation. That builder also doesn't use full `--enable-debug`.

**Comment by hellow554** [comment]
Hey @cuviper. I'm not entirely sure what to do, because I haven't touched this line not even closely. I don't see how "this is my fault" and therefore I don't know exactly what and how to fix it. Maybe you could give me a hint.

**Comment by Centril** [comment]
cc @eddyb @oli-obk @michaelwoerister

**Comment by oli-obk** [comment]
You can replace the offending line with 

```rust
let value = truncate(value, niche.value.size(cx));
```

`truncate` being https://github.com/rust-lang/rust/blob/d20e0002725d2b3efff23c97450806b3517ba804/src/librustc/mir/interpret/mod.rs#L432

**Comment by cuviper** [comment]
@hellow554
>  I don't see how "this is my fault" and therefore I don't know exactly what and how to fix it.

Sorry, I was trying to say it wasn't your fault, but I was too oblique about it. Your change triggered a latent compiler bug, so *someone* has to fix that (here or elsewhere) before you'll pass CI. Looks like you've got it sorted out now though.

**Comment by hellow554** [comment]
I didn't see your comment as an attack, so please don't get me wrong. It was okay, that you said it and I see why. I was just overchallenged, because I never touched that code. But thankfully oli-obk helped me (danke!). I hope this will pass CI now. Thanks again!

File: src/librustc_codegen_llvm/debuginfo/metadata.rs
Comment (eddyb) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: Oh wow I don't know what I thought when I wrote that code, maybe I just forgot the semantics of `<<`, even when it doesn't panic - at least it's not UB!

*Nevermind*, it wasn't me, see https://github.com/rust-lang/rust/pull/55701#discussion_r270227197.
Also, the `assert_eq!` I suggested in https://github.com/rust-lang/rust/commit/da7b6b4b4dcdf9adf25eca12293be029ce9ce24c, during #54004, was removed, which means debuggers may be getting bad values now (if we started allowing `u128` `enum` discriminants).

Comment (eddyb) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: Since you're touching this code, can you add the `assert_eq!(value as u64 as u128, value);` I'm suggesting in https://github.com/rust-lang/rust/pull/55701#discussion_r270227197?

**Review by eddyb — CHANGES_REQUESTED** [review]

**Comment by Centril** [comment]
(let's do the other changes in a follow up... this is included in a pending rollup atm... don't force push!)

**Comment by eddyb** [comment]
Okay, @Centril asked me not to break his rollups so I opened #59509 instead.

@bors r=oli-obk

**Pull request merged**
